components/sup/src/manager/service/hook_runner.rs

```diff
@@ -41,40 +41,28 @@ impl<H> HookRunner<H> where H: Hook + Sync
                      passwd }
     }
 }
-impl<H: Hook + Sync + 'static> IntoFuture for HookRunner<H> {
+
+type LazyClosure<I, E> = Lazy<Box<dyn FnOnce() -> FutureResult<I, E> + Send>, FutureResult<I, E>>;
+
+impl<H: Hook + 'static> IntoFuture for HookRunner<H> {
     type Error = Error;
-    type Future = SpawnedFuture<Self::Item>;
+    type Future = LazyClosure<Self::Item, Self::Error>;
     type Item = (H::ExitValue, Duration);
 
     fn into_future(self) -> Self::Future {
-        let (tx, rx) = oneshot::channel();
-
-        // TODO (CM): Consider using a short abbreviation for the hook
-        // name in the thread name (e.g. "HC" for "health_check", "I"
-        // for "init", etc.
-
         // TODO (CM): May want to consider adding a configurable
         // timeout to how long this hook is allowed to run.
-        let handle_result =
-            thread::Builder::new().name(format!("{}-{}", H::file_name(), self.service_group))
-                                  .spawn(move || {
-                                      // _timer is for Prometheus metrics, but we also want
-                                      // the runtime for other purposes. Unfortunately,
-                                      // we're not able to use the same timer for both :(
-                                      let _timer = hook_timer(H::file_name());
-                                      let start = Instant::now();
-                                      let exit_value = self.hook.run(&self.service_group,
-                                                                     &self.pkg,
-                                                                     self.passwd.as_ref());
-                                      let run_time = start.elapsed();
-                                      tx.send((exit_value, run_time))
-                                        .expect("Couldn't send oneshot signal from HookRunner: \
-                                                 receiver went away");
-                                  });
-
-        match handle_result {
-            Ok(_handle) => rx.into(),
-            Err(io_err) => io_err.into(),
-        }
+        future::lazy(Box::new(move || {
+                         // _timer is for Prometheus metrics, but we also want
+                         // the runtime for other purposes. Unfortunately,
+                         // we're not able to use the same timer for both :(
+                         let _timer = hook_timer(H::file_name());
+                         let start = Instant::now();
+                         let exit_value =
+                             self.hook
+                                 .run(&self.service_group, &self.pkg, self.passwd.as_ref());
```

Comment:
The problem here is that a long-running hook will gum up the executor, because this future will only yield once the hook is done running. Though the previous formulation is more verbose, it has the advantage of keeping the heavy lifting off to the side in another thread; the oneshot channel is the only real future bit there, and polling it is a very quick operation.

See https://docs.rs/futures/0.1.27/futures/future/trait.Future.html#runtime-characteristics for details.

The closure that `future::lazy` takes is just run normally (see [here](https://docs.rs/futures/0.1.27/src/futures/future/lazy.rs.html#64)), so there's no opportunity to break out of it and come back later if it's taking a long time to run.

The comment on Line 4 above points the way forward to a cleaner implementation.